Port to python3 and add functionality

Dimi20cen opened the pull request:
- Fixed issue with adjusting pygame window to screensize
- Added mouseclicking for changing lanes
- Added pause functionality, either with clicking on pause button or by pressing space
- Added homebutton, during scorescreen

@chimosky @sourabhaa

ayushnawal committed:
update sugargame to v1.3

ayushnawal committed:
Port to Python 3

ayushnawal committed:
changes made as requested

sourabhaa committed:
fix: fixed sound icon

sourabhaa committed:
activity works but there's issue with pygame initialisation

sourabhaa committed:
refactor: fixed spacing after the assignmet

quozl committed:
Allow execution of files

Two Python files did not have executable permissions.

Shouldn't cause any symptom, but was irritating.

Signed-off-by: James Cameron <quozl@laptop.org>

quozl committed:
Adopt conventions

- camel case the game class name,

- rename source file to match lowercase class name,

- rename instance variable to match Sugargame test activity

Signed-off-by: James Cameron <quozl@laptop.org>

quozl committed:
Remove show_all in toolbar

Sugargame test activity does not do this.

Was also found in other activities, such as Bridge and Jumble, it
changes the order of creation of the X11 windows.

The earlier port of thie activity to Sugargame 1.3 has not properly
followed the new API and test activity.

Signed-off-by: James Cameron <quozl@laptop.org>

quozl committed:
Fix path to score file

EPERM is reported when the activity is run under Sugar when the source
code is in /usr/share/sugar/activities

quozl committed:
Add missing get_preview method

quozl committed:
Show the sound button

quozl committed:
Port to Sugargame 1.3 API

- name the run method the same as the example program,

- use a running flag rather than a crashed flag; crashed could be
  misunderstood as referring to the cars, and running is how the
  example program is phrased,

- change initialisation of the display,

- properly exit the event loop if the stop button is pressed,

- handle all events in the message queue in one place.

Tested on Ubuntu 20.04 with Sugar.  Mostly works.

Signed-off-by: James Cameron <quozl@laptop.org>

chimosky commented on activity.py:
```diff
@@ -57,16 +63,20 @@ def build_toolbar(self):
         stop_button.show()
+        stop_button.connect('clicked', self._stop_cb)
 
-        self.show_all()
```
Why remove this?

chimosky commented on activity/activity.info:
```diff
@@ -1,9 +1,9 @@
-exec = sugar-activity activity.Activity
-license = GPLv3
+exec = sugar-activity3 activity.Activity
+license = GPLv3+
```
Does the fork contain these changes too?

chimosky reviewed (commented):
If these changes are from Sourabha's PR then that PR contains 13 commits which it looks like you've merged all to one without retaining the commit authors.

If you intend to continue the work on that PR then fetch the PR and then continue on top of the latest commit.

chimosky reviewed (changes requested):
Could you keep flake 8 changes to another PR?

chimosky commented on elements.py:
```diff
@@ -1,83 +1,53 @@
                 game.objectlist.remove(self)
 
-        if (game.move == 1):
+        if game.move:
```
The check seems valid so why change it?

chimosky commented on elements.py:
```diff
@@ -1,83 +1,53 @@
-            550 + game.rightcar.get_height() / 2))
+        element_rect = self.element.get_rect(center=(self.x + self.element.get_width() / 2, self.y + self.element.get_height() / 2))
+        car1_rect = game.leftcar.get_rect(center=(game.leftcar_x + game.leftcar.get_width() / 2, 550 + game.leftcar.get_height() / 2))
+        car2_rect = game.rightcar.get_rect(center=(game.rightcar_x + game.rightcar.get_width() / 2, 550 + game.rightcar.get_height() / 2))
```
This isn't pep8 compliant so you can leave the old lines and then if you want to change it then do that in another flake8 related PR.

chimosky commented on game.py:
```diff
@@ -20,21 +20,70 @@
 from gi.repository import Gtk
 import pygame
 from gettext import gettext as _
+from pygame.locals import *
```
Would be great if you can avoid wildcard imports.

chimosky commented on game.py:
```diff
@@ -20,21 +20,70 @@
 from welcomescreen import Welcomescreen
 from scorescreen import Scorescreen
 
+os.environ['SUGAR_BUNDLE_ID'] = 'org.sugarlabs.2cars'
```
Why are you doing this?

chimosky commented on game.py:
```diff
@@ -51,203 +100,176 @@ def initialize(self):
-            # Gtk events
-            while Gtk.events_pending():
-                Gtk.main_iteration()
             if not self.running:
```
Why remove this?

chimosky commented on game.py:
The changes in this file need to be explained properly, the commit message " Now works with python3 and opens widows with sugar-activity3" isn't helpful. 

Please make your commit messages more helpful, see [making commits](https://github.com/sugarlabs/sugar-docs/blob/master/src/contributing.md#making-commits)

chimosky commented on scorescreen.py:
```diff
@@ -20,89 +20,95 @@
+    WHITE = (255, 255, 255)
+    SCORESCREEN_IMAGE_PATH = "assets/scorescreen.png"
+    RESTART_IMAGE_PATH = "assets/restart.png"
+    FONT_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "fonts/Arimo.ttf")
```
As these are supposed to be constants, why not make them global?

chimosky commented on sugargame/event.py:
This isn't maintained by the activity, any changes to sugargame should be made in the [sugargame repo](https://github.com/sugarlabs/sugargame) and tested with the sugargame activity and any activity that needs it.

The print statement tells me it was probably for debug and you forgot to remove it.

Dimi20cen replied:
I tried to configure the event.py file, because I had some issues with key events when testing with `sugar-activity3 .`. And the print was debugging. I now replaced the sugargame with the 1.3 version

chimosky commented on welcomescreen.py:
```diff
@@ -26,26 +26,33 @@
+    PLAY_IMAGE_PATH = "assets/play.png"
+    RULES_IMAGE_PATH = "assets/rules.png"
+    BUTTON_IMAGE_PATH = "assets/button.png"
+
```
Please make these global as they're constants.

chimosky commented on game.py:
```diff
@@ -111,24 +114,29 @@ def initialize(self):
-        self.game_display = screen
         self.convert_images()
 
+        self.WIN = pygame.display.get_surface()
```
Why the naming convention change?

chimosky commented on game.py:
```diff
@@ -37,251 +37,301 @@
+RED_CIRCLE_PATH = os.path.join("assets/redcircle.png")
+BLUE_CIRCLE_PATH = os.path.join("assets/bluecircle.png")
+RED_SQUARE_PATH = os.path.join("assets/redsquare.png")
+BLUE_SQUARE_PATH = os.path.join("assets/bluesquare.png")
```
If you're using `os.path.join` then you can't pass the relative path yourself as there will be no need to use it if you did.

Dimi20cen replied:
Fixed

chimosky commented on game.py:
```diff
@@ -37,251 +37,301 @@
         pygame.font.init()
+        pygame.mixer.init()
         self.load_assets()
+        print("Loading game...")
```
Another debug?

Dimi20cen replied:
I thought it was a fun terminal output, but I understand terminal isn't  the place to have fun, removed it

chimosky commented on game.py:
```diff
@@ -37,251 +37,301 @@
+        self.red_cricle_image = pygame.image.load("assets/redcircle.png")
+        self.blue_circle_image = pygame.image.load("assets/bluecircle.png")
+        self.red_square_image = pygame.image.load("assets/redsquare.png")
+        self.blue_square_image = pygame.image.load("assets/bluesquare.png")
```
You created constants for this above so why not use them?

Dimi20cen replied:
You are correct, fixed it

chimosky commented on game.py:
```diff
@@ -37,251 +37,301 @@
+        self.RED_CIRCLE = pygame.transform.scale(self.red_cricle_image, (self.obj_dimension, self.obj_dimension))
+        self.BLUE_CIRCLE = pygame.transform.scale(self.blue_circle_image, (self.obj_dimension, self.obj_dimension))
+        self.RED_SQUARE = pygame.transform.scale(self.red_square_image, (self.obj_dimension, self.obj_dimension))
+        self.BLUE_SQUARE = pygame.transform.scale(self.blue_square_image, (self.obj_dimension, self.obj_dimension))
```
This isn't a good way to go about this, if they're to be globals then create then when you should, if they're not then name them accordingly.

Dimi20cen replied:
fixed

chimosky commented on game.py:
```diff
@@ -37,251 +37,301 @@
 
     def initialize(self):
+        self.screen_width, self.screen_height = pygame.display.get_surface().get_size()
+        self.MIDDLE_OF_SCREEN_X = self.screen_width / 2
```
Name this properly if it's only used here.

This and every other one like it.

Dimi20cen replied:
From past commit

Dimi20cen replied:
Will do

chimosky commented on game.py:
```diff
@@ -37,251 +37,301 @@
+        LEFT_CAR_LEFT_BOUND = self.lane_1 - self.car_center  # How far the car can go left
+        LEFT_CAR_RIGHT_BOUND = self.lane_2 - self.car_center   # How far the car can go right
+        RIGHT_CAR_LEFT_BOUND = self.lane_3 - self.car_center   # How far the car can go left
+        RIGHT_CAR_RIGHT_BOUND = self.lane_4 - self.car_center   # How far the car can go right
```
If these are to be used only here then name them accordingly.

Dimi20cen replied:
Will do

chimosky commented on game.py:
```diff
@@ -37,251 +37,301 @@
+    # pygame.display.set_mode((500, 350))
+    # pygame.display.set_mode((2024, 1468))
+    # pygame.display.set_mode((1920, 1080))
+    # pygame.display.set_mode((491, 768))
```
You can get rid of the commented code.

Dimi20cen replied:
Back then I mainly used `sugar-activity3 .` to test. They were handy for testing. Removed them

chimosky commented on scorescreen.py:
```diff
@@ -72,55 +82,62 @@ def write_new_max_score(self, new_max_score):
+        # Just a window exception check condition??????????????????????????????
+        # event1 = pygame.event.get()
+        # if event1.type is pygame.QUIT:
+        #     crashed = True
```
Delete the commented out code.

Dimi20cen replied:
FIxed

chimosky commented on scorescreen.py:
```diff
@@ -72,55 +82,62 @@ def write_new_max_score(self, new_max_score):
-        if event1.type is pygame.QUIT:
-            crashed = True
+        # if crashed is True:
+        #     pygame.quit()
```
This too.

chimosky commented on elements.py:
```diff
@@ -21,15 +21,17 @@ def display(self, game):
+        # changing it will affect perceived speed of the cars
         if game.move:
-            self.y_to_draw += 5
+            self.y_to_draw += game.background_height * 0.0065
```
How'd `0.0065` come about?

Dimi20cen replied:
Around this value the speed of the cars feels ideal. Do you suggest something else?

chimosky replied:
Okay.

chimosky commented on game.py:
```diff
@@ -241,8 +250,8 @@ def handle_collision(self):
-                self.collision = 1
-        if self.collision == 1:
+                self.collision = True
+        if self.collision is True or self.coin_miss is True:
```
This can be `if self.collision or self.coin_miss:`

Dimi20cen replied:
Fixed

chimosky commented on game.py:
```diff
@@ -255,13 +264,18 @@ def create_object_list(self):
                         break
+                for x in range(len(self.objectlist)):
+                    print(self.objectlist[x])
+                # print(*self.objectlist, sep = "\n ")
```
Another debug help?

Dimi20cen replied:
Yep, fixed

chimosky commented on game.py:
```diff
@@ -75,41 +80,42 @@
-        self.leftcar_image.set_colorkey((0, 0, 0))
         self.rightcar_image = pygame.image.load(BLUECAR_IMAGE_PATH)
-        self.rightcar_image.set_colorkey((0, 0, 0))
+        # self.rightcar_image.set_colorkey((0, 0, 0))
```
Delete if not needed.

Dimi20cen replied:
yep

chimosky commented on activity.py:
"Changed the activity.py file" doesn't tell what you did, please write better commit messages.

Dimi20cen replied:
Sorry for that

chimosky commented on elements.py:
"Small final changes for working version" isn't helpful, please rewrite the commit message to tell what changed and where necessary why.

Dimi20cen replied:
yep

chimosky commented:
Tested 27d2652, arrow keys no longer work and only mouse buttons work.

Dimi20cen commented:
In my environments it works just fine, seems to be the same issue I faced with flappy's toolbar not showing

jriyyya commented:
Tested [27d265](https://github.com/sugarlabs/2-cars-activity/commit/27d265220397cf77988b617c8b98ce8f5a4fdfa5), Arrow keys are working for me

Dimi20cen commented:
> If you intend to continue the work on that PR then fetch the PR and then continue on top of the latest commit.

Should I do that now? I think it's possible

>Could you keep flake 8 changes to another PR?

I can't figure out how to approach flake8 for python2, can you help me? I did an attempt #13

quozl commented:
I can't tell if this set of changes is based on the earlier pull requests by sourabhaa or ayushnawal, or if it discards all that work we(-all) did in favour of doing it over again.  Can you explain if you used the earlier work and just forgot to use "git commit --author", or if you just didn't notice it?

Looking at the flake8 fixes for python2 seems unhelpful unless you plan to rebase your branch on top of them.  Is that what you plan to do?

Dimi20cen commented:
> Can you explain if you used the earlier work and just forgot to use "git commit --author", or if you just didn't notice it?

The changes in my first commit were based on the sourabhaa's repo, I honestly didn't know  "git commit --author" was a thing. That's why in my commit message I wrote, "Using the already ported to Python3 fork from Sourabhaa's repo.".

> Looking at the flake8 fixes for python2 seems unhelpful unless you plan to rebase your branch on top of them. Is that what you plan to do?

No, I don't think I want to rebase my branch on top of them. I thought it might be helpful for some other reason, since it was suggested.

quozl commented:
> The changes in my first commit were based on the sourabhaa's repo, I honestly didn't know "git commit --author" was a thing. That's why in my commit message I wrote, "Using the already ported to Python3 fork from Sourabhaa's repo.".

We usually use git to attribute developer work, and the format used by git for author ensures that the developer is identified properly.  Sourabhaa is a GitHub username, and that isn't stored in git.  Perhaps you could rebase your branch on their branch?  That will ensure their commits are present.

> No, I don't think I want to rebase my branch on top of them. I thought it might be helpful for some other reason, since it was suggested.

I expect Ibiam did this to keep flake8 work separate from more substantial changes.  It is much harder to review a pull request or especially a commit where there is obvious confusion about what the commit is for; and mixing flake8 work with substantial change is an example of that.

In our guide to reviewers in the contributing guide, we have summarised this as a style issue;

> * [ ] is the changed code consistent in style with the existing code, see also [coding standards](desktop-activity.md#coding-standards), (on the other hand, expect flake8 changes to be in separate commits),

Perhaps you need to do some reviews yourself, to better understand the process.  Have you done any?

Dimi20cen committed:
Add .flake8 and fix flakes

Dimi20cen committed:
Refactor activity

Aim is to improve readability mainly by
- Use more descriptive naming for variables and functions
- Breaking up large functions
- Adding global variables

Other changes
- Incorporate score_path inside scorescreen
- Fill space that was unused by pygame with black color

Dimi20cen committed:
Make pygame display adjust to screen size

Make the following adjust to screen size
- Background size
- Squares' and circles' size
- Cars' size
- Font size
- Perceived speed of cars

Dimi20cen committed:
Add lane change with mouse clicking

Dimi20cen committed:
Add home button in scorescreen

- Add home button image
- Use the same code that's used for restart button

Dimi20cen committed:
Add pause functionality

When pause button is clicked or space button is pressed game is paused and screen darkens

Changes
- Add pause button image
- Display button at the top left
- Add handling for when pause is clicked
- Add handling for when space key is press
- Add method for when pause button or space key is pressed

quozl commented:
Thanks.  Reviewed to 988afb2.

Dimi20cen committed:
Update sugargame

This change also fixes the following Traceback that was caused when using the activity with newer pygame versions
-  Traceback (most recent call last):
  File "/home/dim/Activities/2-cars-activity/sugargame/event.py", line 128, in _keydown_cb
    return self._keyevent(widget, event, pygame.KEYDOWN)
  File "/home/dim/Activities/2-cars-activity/sugargame/event.py", line 169, in _keyevent
    mod = self._keymods()
  File "/home/dim/Activities/2-cars-activity/sugargame/event.py", line 145, in _keymods
    mod |= self.__keystate[key_val] and mod_val
IndexError: list index out of range

Dimi20cen commented:
Arrow keys should be working with the updated version of sugargame. @chimosky

chimosky commented:
Please drop 52ed8e0 as that's not related to this PR, you can open another PR for it.

I haven't looked at #13 yet but it should be there instead of here.

The refactor in ca712d3 isn't also needed here as it's not related to this PR, another PR for that would be helpful.

chimosky commented on game.py:
```diff
@@ -229,21 +265,21 @@ def handle_collision(self):
     def create_object_list(self):
         if self.tick_counter == 10 or self.tick_counter == 35:
             self.soon_to_draw = Element(self)
```
As this is only used in this function, there's no need to make it an instance variable. 

A better change would've been to rename `self.soon_to_draw` to fit the other uses.

Dimi20cen replied:
Done, replace it with `pending_element`

chimosky commented on game.py:
```diff
@@ -259,21 +295,25 @@ def update_car_angles(self):
+        left_car_right_bound = self.lane_2 - self.car_center
+        right_car_left_bound = self.lane_3 - self.car_center
+        right_car_right_bound = self.lane_4 - self.car_center
         # Left car
```
Would be great to have a new line before the comment, helps make the code less clustered.

Dimi20cen replied:
done

chimosky commented on scorescreen.py:
```diff
@@ -106,27 +108,52 @@ def run(self, game, scores):
+            restart_rect = self.restart.get_rect(
+                center=(restart_x, (game.screen_height // 2)))
+            if restart_rect.collidepoint(mos_x, mos_y):
+
```
Why the new line before the first  assignment?

chimosky commented on welcomescreen.py:
```diff
@@ -80,31 +84,55 @@ def run(self, game):
             # Show Welcome screen
-            if ruleflag == 0:
-                game.screen.blit(self.welcomescreen, (350, 0))
+            if (ruleflag == 0):
```
No need for the parentheses.

Dimi20cen replied:
done

chimosky commented on game.py:
```diff
@@ -242,17 +263,66 @@ def handle_input(self, event):
+                Gtk.main_iteration()
+            for event in pygame.event.get():
+                if event.type == pygame.QUIT:
+                    pygame.quit()
+                    sys.exit()
```
We don't use a `pygame.quit` and `sys.quit` on event type `pygame.QUIT`, you can just use a return.

See the test activity in sugargame.

Dimi20cen replied:
I replaced it with `self.running = False` for code consistency

chimosky commented on game.py:
```diff
@@ -242,17 +263,66 @@ def handle_input(self, event):
+                         self.background_width * 0.4),
+                        (self.screen_height * 0.065)))
+
+                if (pause_rect.collidepoint(mouse_x, mouse_y) and
+                   event.type == pygame.MOUSEBUTTONUP and
+                   event.button == 1):
```
No need for the outer parentheses.

Dimi20cen replied:
Without the outer parentheses, I would get `SyntaxError: invalid syntax`

chimosky replied:
Use `\` to indicate that the line doesn't end there.

Dimi20cen commented:
> Please drop https://github.com/sugarlabs/2-cars-activity/commit/52ed8e0bf1e8f1136c90d82403881741afb54147 as that's not related to this PR, you can open another PR for it.
I haven't looked at https://github.com/sugarlabs/2-cars-activity/pull/13 yet but it should be there instead of here.

The refactor in https://github.com/sugarlabs/2-cars-activity/commit/ca712d36821540d9a8bb2abe5dc5acc6a708bf0c isn't also needed here as it's not related to this PR, another PR for that would be helpful.

I read that in the guide, but in this specific case, I fixed flakes and refactored the code and then on top of those I added new features. So my features were build on top of the flake fixes and the refactoring. Wouldn't the only other choice be to make a PR for the flakes, then a PR on top of it for the refactoring and then another PR on top of them for the added features? Wouldn't that create too much complexity? I also had in mind that in the case of this small project that would be to confusing.

Dimi20cen committed:
Improve code consistency

Mainly
- Add empty lines above comments
- Rename self.soon_to_draw to pending_element
- Set line to execute when `if event.type == pygame.QUIT:` to `self.running = False`

chimosky commented:
> I read that in the guide, but in this specific case, I fixed flakes and refactored the code and then on top of those I added new features. So my features were build on top of the flake fixes and the refactoring. Wouldn't the only other choice be to make a PR for the flakes, then a PR on top of it for the refactoring and then another PR on top of them for the added features? Wouldn't that create too much complexity? I also had in mind that in the case of this small project that would be to confusing.

My bad, I should've explained earlier, this PR should be for *Port to Python 3*, any new functionality should be in another PR, same goes for any other change that's not a *Port to Python 3*.
That means you can base a different branch on the necessary changes and keep each branch to changes that needs to be there, so flake8 on a different branch and added functionalities on a different branch.
